### PR TITLE
[WIP] Added support for relations in logAttributes

### PR DESCRIPTION
**Pull request by vanbrabantf**
### Changed
  - Added support for relations in logAttributes (#90)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
         "php" : "^7.0",
         "illuminate/config": "~5.1.24|~5.2.0|~5.3.0|~5.4.0",
         "illuminate/database": "~5.1.24|~5.2.6|~5.3.0|~5.4.0",
-        "illuminate/support": "~5.1.24|~5.2.0|~5.3.0|~5.4.0",
+        "illuminate/support": "~5.3.0|~5.4.0",
```
Comment (freekmurze) [inline composer.json]: Why did you remove the older versions?

File: src/Traits/DetectsChanges.php [hunk 2]
```diff
@@ -24,13 +24,21 @@ protected static function bootDetectsChanges()
...
     public function attributesToBeLogged(): array
     {
-        if (! isset(static::$logAttributes)) {
+        if (!isset(static::$logAttributes)) {
             return [];
         }
 
-        return static::$logAttributes;
+        return collect(static::$logAttributes)->map(
+            function ($value) {
```
Comment (freekmurze) [inline src/Traits/DetectsChanges.php]: add a string typehint here

File: src/Traits/DetectsChanges.php [hunk 3]
```diff
@@ -24,13 +24,21 @@ protected static function bootDetectsChanges()
...
+                if (strpos($value, '.') != 0) {
```
Comment (freekmurze) [inline src/Traits/DetectsChanges.php]: You can replace this with the more readable `starts_with` function.

File: src/Traits/DetectsChanges.php [hunk 4]
```diff
@@ -50,6 +58,20 @@ public function attributeValuesToBeLogged(string $processingEvent): array
 
     public static function logChanges(Model $model): array
     {
-        return collect($model)->only($model->attributesToBeLogged())->toArray();
+        return collect($model->attributesToBeLogged())->mapWithKeys(
+            function ($value) use ($model) {
+                if (is_iterable($value)) {
```
Comment (freekmurze) [inline src/Traits/DetectsChanges.php]: `is_iterable` only exists in PHP 7.1, I'm not ready to drop PHP 7.0 support just yet.

File: src/Traits/DetectsChanges.php [hunk 5]
```diff
@@ -50,6 +58,20 @@ public function attributeValuesToBeLogged(string $processingEvent): array
...
+                    foreach ($value as $methodCall) {
+                        $model = $model->$methodCall;
+                    }
+
+                    return [implode('.', $value) => $model];
+                } else {
```
Comment (freekmurze) [inline src/Traits/DetectsChanges.php]: Generally I try to avoid `else` statements. Since you're using an early return in the `if` branch, you can just not use and `else`statement and put the second `return` after closing the `if` branch.

File: tests/TestCase.php [hunk 6]
```diff
@@ -85,12 +85,17 @@ public function getTempDirectory(): string
...
         collect($tableNames)->each(function (string $tableName) {
-            $this->app['db']->connection()->getSchemaBuilder()->create($tableName, function (Blueprint $table) {
+            $this->app['db']->connection()->getSchemaBuilder()->create($tableName, function (Blueprint $table) use ($tableName){
                 $table->increments('id');
                 $table->string('name')->nullable();
                 $table->string('text')->nullable();
                 $table->timestamps();
                 $table->softDeletes();
+
+                if ($tableName === 'articles') {;
```
Comment (freekmurze) [inline tests/TestCase.php]: `;` can be removed on this line

Comment (freekmurze) [inline tests/TestCase.php]: Instead of putting an if here, just create a second  `$this->app['db']->connection()->getSchemaBuilder()->create` statement where you create the articles table.

File: composer.json (hunk 1 above)
Comment (vanbrabantf) [inline composer.json]: `mapWithKeys` is used here: https://github.com/spatie/laravel-activitylog/pull/122/files#diff-6bae372eb15ba8af41415ab509d0982bR61

File: src/Traits/DetectsChanges.php (hunk 3 above)
Comment (vanbrabantf) [inline src/Traits/DetectsChanges.php]: This checks if there are dots in the string for example `example.deeper`.
`starts_with` would not catch these right?

File: composer.json (hunk 1 above)
Comment (freekmurze) [inline composer.json]: Ok, drop support for older versions of `illuminate/config` and `illuminate/support` as well

File: src/Traits/DetectsChanges.php (hunk 3 above)
Comment (freekmurze) [inline src/Traits/DetectsChanges.php]: You're right. I read the code wrong. But you could use `str_contains` instead.

**Comment by freekmurze** [comment]
Thank you for your work on this.

I pulled this in a separate branch so I can work a bit on this myself.

**Comment by LukeTowers** [comment]
@freekmurze @vanbrabantf why was support dropped for Laravel 5.1 LTS? I don't see any reason that it wouldn't work with 5.1 besides the fact that 5.1 was removed from `composer.json`.

**Comment by freekmurze** [comment]
Don't know for sure anymore, but I'm thinking we did change it because a signature chance in some of the collection methods.

Clone the package, add support for L5.1, require the lower dependencies and run the tests to see what actually is breaking.

**Comment by LukeTowers** [comment]
Looks like it was the `mapWithKeys` collection helper method, added in Laravel 5.3. We could easily substitute that by using the `each` collection helper instead. @freekmurze would you accept that in a PR?

**Comment by LukeTowers** [comment]
@freekmurze Made the necessary changes and created #184 to get them merged in. Let me know if I need to make any other changes to that.
